### ISSUE TITLE
Implementation architecture

### ISSUE THREAD

**Issue by mielvds** [SPARQL-Anything/sparql.anything#211]
Hi all!

I was wondering if there is some info on how SPARQL Anything is implemented? I gather it's mostly Jena, but a diagram of all the components would be helpful in case one were to make improvements.

**Comment by enridaga** [comment]
Agreed, we need to develop some technical documentation on the implementation, e.g. how to develop new functions or new adapters.

Key places to look at are:

- The command line interface (to see the end-to-end process): [SPARQLAnything.java](https://github.com/SPARQL-Anything/sparql.anything/blob/9917eac65ed2b4e78a8c57b99d13caa695c11229/sparql-anything-cli/src/main/java/com/github/sparqlanything/cli/SPARQLAnything.java)
- The query executor, who handles parameters and runs the triplifiers: [FacadeXOpExecutor.java](https://github.com/SPARQL-Anything/sparql.anything/blob/9917eac65ed2b4e78a8c57b99d13caa695c11229/sparql-anything-engine/src/main/java/com/github/sparqlanything/engine/FacadeXOpExecutor.java)
- A register of available triplifiers and extension functions: [FacadeX.java](https://github.com/SPARQL-Anything/sparql.anything/blob/9917eac65ed2b4e78a8c57b99d13caa695c11229/sparql-anything-engine/src/main/java/com/github/sparqlanything/engine/FacadeX.java)
- The triplifiers themselves, for example [CSV](https://github.com/SPARQL-Anything/sparql.anything/blob/9917eac65ed2b4e78a8c57b99d13caa695c11229/sparql-anything-csv/src/main/java/com/github/sparqlanything/csv/CSVTriplifier.java), [JSON](https://github.com/SPARQL-Anything/sparql.anything/blob/9917eac65ed2b4e78a8c57b99d13caa695c11229/sparql-anything-json/src/main/java/com/github/sparqlanything/json/JSONTriplifier.java), or [Markdown](https://github.com/SPARQL-Anything/sparql.anything/blob/9917eac65ed2b4e78a8c57b99d13caa695c11229/sparql-anything-markdown/src/main/java/com/github/sparqlanything/markdown/MARKDOWNTriplifier.java)
- The FX builder, which is used by the triplifiers, the default is [TripleFilteringFacadeXGraphBuilder.java](https://github.com/SPARQL-Anything/sparql.anything/blob/9917eac65ed2b4e78a8c57b99d13caa695c11229/sparql-anything-model/src/main/java/com/github/sparqlanything/model/TripleFilteringFacadeXGraphBuilder.java)

Hope this helps

**Comment by justin2004** [comment]
I know this doesn't address the absence of documentation but my general approach for getting the lay of the land is to find a line(s) of code I basically understand (in isolation) and then insert a break point there or add a "print stack trace" statement to see how execution arrives there.
That is usually enough to orient me and then my own interest in the project has to take me the rest of the way.
